### PR TITLE
Profiler test render tracking

### PR DESCRIPTION
**Pull request by phryneas**
@jerelmiller can you take a look?

### Checklist:

- [ ] If this PR contains changes to the library itself (not necessary for e.g. docs updates), please include a changeset (see [CONTRIBUTING.md](https://github.com/apollographql/apollo-client/blob/main/CONTRIBUTING.md#changesets))
- [ ] If this PR is a new feature, please reference an issue where a consensus about the design was reached (not necessary for small changes)
- [ ] Make sure all of the significant new logic is covered by tests

### CODE REVIEW

File: src/react/hooks/__tests__/useLoadableQuery.test.tsx [hunk 1]
```diff
@@ -281,13 +283,22 @@ it("loads a query with variables and suspends by passing variables to the loadQu
...
-  expect(App).toHaveRenderedTimes(2);
+  {
+    const { renderedComponents } = await App.takeRender();
+    expect(renderedComponents).toStrictEqual(["App", "SuspenseFallback"]);
+  }
+
+  {
+    const { renderedComponents } = await App.takeRender();
+    expect(renderedComponents).toStrictEqual(["UseReadQuery"]);
+  }
```
Comment (phryneas) [inline src/react/hooks/__tests__/useLoadableQuery.test.tsx:301]: For components to appear here, they have to either be wrapped in `profile(`, or they need to call the `useTrackComponentRender`  hook.

File: src/testing/internal/profile/profile.tsx [hunk 2]
```diff
@@ -170,13 +188,20 @@ export function profile<
...
+
   let iteratorPosition = 0;
   const Profiled: ProfiledComponent<Props, Snapshot> = Object.assign(
-    (props: Props) => (
-      <React.Profiler id="test" onRender={profilerOnRender}>
-        <Component {...(props as any)} />
-      </React.Profiler>
-    ),
+    (props: Props) => {
+      const parentContext = React.useContext(ProfilerContext);
```
Comment (phryneas) [inline src/testing/internal/profile/profile.tsx:196]: renders are always tracked only in the outmost profiled component.

File: src/testing/internal/profile/profile.tsx [hunk 3]
```diff
@@ -339,3 +366,64 @@ export function profileHook<ReturnValue extends ValidSnapshot, Props>(
...
+  while (proto && proto !== Object) {
+    if (proto === React.Component) return true;
+    proto = Object.getPrototypeOf(proto);
+  }
+  return false;
+}
+
+function getCurrentComponentName() {
+  const owner: React.ComponentType | undefined = (React as any)
+    .__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED?.ReactCurrentOwner
```
Comment (phryneas) [inline src/testing/internal/profile/profile.tsx:383]: 🙈

**Review by jerelmiller — APPROVED** [review]
I may tweak a few things in the base branch, but this is a great start. Thanks!

**Comment by jerelmiller** [comment]
Will fixup failures in the base branch

**Pull request merged**